python_modules/dagster/dagster/_config/structured_config.py

```diff
@@ -15,6 +20,66 @@ class Config(BaseModel):
     """
 
 
+class Resource(
+    ResourceDefinition,
+    Config,
+):
+    """
+    Base class for Dagster resources that utilize structured config.
+
+    This class is a subclass of both :py:class:`ResourceDefinition` and :py:class:`Config`, and
+    provides a default implementation of the resource_fn that returns the resource itself.
+
+    Example:
+
+    .. code-block:: python
+
+        class WriterResource(Resource):
+            prefix: str
+
+            def output(self, text: str) -> None:
+                print(f"{self.prefix}{text}")
+
+    """
+
+    def __init__(self, **data: Any):
+        schema = infer_schema_from_config_class(self.__class__)
+
+        inner_resource_def = ResourceDefinition(self.resource_function, schema)
+        configured_resource_def = inner_resource_def.configured(
+            config_dictionary_from_values(
+                data,
+                schema,
+            ),
+        )
+
+        Config.__init__(self, **data)
+        ResourceDefinition.__init__(
+            self,
+            resource_fn=self.resource_function,
+            # mypy worries that configured_resource_def could be self, which
+            # has not had config_schema initialized yet, but we know that's not the case
+            config_schema=configured_resource_def.config_schema,  # type: ignore[attr-defined]
+            description=self.__doc__,
+        )
+
+    def __setattr__(self, name: str, value: Any):
+        # This is a hack to allow us to set attributes on the class that are not part of the
+        # config schema. Pydantic will normally raise an error if you try to set an attribute
+        # that is not part of the schema.
+
+        if name.startswith("_"):
+            object.__setattr__(self, name, value)
+            return
+
+        return super().__setattr__(name, value)
+
```

Comment:
Can you add tests to see if `functools.cached_property` and our own `cached_method` works? I think that will be super useful pattern.

Reply:
Also add tests that test the core capability.

Reply:
Added some more tests, including testing the functionality of caching using `@cached_property` and `@cached_method`.